[flow: servicemonitors] panic: runtime error: invalid memory address or nil pointer dereference

### What's wrong?

Flow Agent with "prometheus.operator.servicemonitors" configured crashes at startup when relablings is configured on ServiceMonitor object. i.e. the Loki servicemonitor object installed with the loki helm chart.

Removing relablings configuration enables the agent to start successfully.

### Steps to reproduce

1. Start a grafana-agent in kubernetes cluster with a ServiceMonitor object with relablings configuration defined.
2. Agent crashes with logs provided further down in this issue

### System information

Linux

### Software version

next

### Configuration

```text
prometheus.operator.servicemonitors "servicemonitors" {
            forward_to = [prometheus.remote_write.mimir.receiver]
        }        

        prometheus.remote_write "mimir" {
          endpoint {
            url = "http://mimir:9009/api/v1/push"
          }

        }
```


### Logs

```text
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0xbbb308]

goroutine 417 [running]:
github.com/grafana/regexp.(*Regexp).FindStringSubmatchIndex(0x0, {0xc001069638?, 0x8?})
	/go/pkg/mod/github.com/grafana/regexp@v0.0.0-20221122212121-6b5c0a4cb7fd/regexp.go:1065 +0x28
github.com/prometheus/prometheus/model/relabel.relabel({0xc0024e7000?, 0x49, 0x80}, 0xc000377880, 0xc0024a1850)
	/go/pkg/mod/github.com/grafana/prometheus@v1.8.2-0.20230511165250-22c61d1811b2/model/relabel/relabel.go:250 +0x758
github.com/prometheus/prometheus/model/relabel.Process({0xc0024e7000, 0x43, 0x80}, {0xc001630300, 0x10, 0x0?})
	/go/pkg/mod/github.com/grafana/prometheus@v1.8.2-0.20230511165250-22c61d1811b2/model/relabel/relabel.go:211 +0x226
github.com/prometheus/prometheus/scrape.PopulateLabels({0xc0024e8800, 0x3e, 0x3e}, 0xc000c9ae00, 0x0)
	/go/pkg/mod/github.com/grafana/prometheus@v1.8.2-0.20230511165250-22c61d1811b2/scrape/target.go:376 +0x58a
github.com/prometheus/prometheus/scrape.TargetsFromGroup(0xc002259ef0, 0xc000c9ae00, 0xd1?)
	/go/pkg/mod/github.com/grafana/prometheus@v1.8.2-0.20230511165250-22c61d1811b2/scrape/target.go:506 +0x2b1
github.com/prometheus/prometheus/scrape.(*scrapePool).Sync(0xc001730b40, {0xc001630900, 0x9, 0xc0018c04e0?})
	/go/pkg/mod/github.com/grafana/prometheus@v1.8.2-0.20230511165250-22c61d1811b2/scrape/scrape.go:495 +0x210
github.com/prometheus/prometheus/scrape.(*Manager).reload.func1(0xc001678f60?, {0xc001630900?, 0x1ad8966?, 0xc0018c0480?})
	/go/pkg/mod/github.com/grafana/prometheus@v1.8.2-0.20230511165250-22c61d1811b2/scrape/manager.go:228 +0x27
created by github.com/prometheus/prometheus/scrape.(*Manager).reload
	/go/pkg/mod/github.com/grafana/prometheus@v1.8.2-0.20230511165250-22c61d1811b2/scrape/manager.go:227 +0x515
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -99,6 +99,8 @@ Main (unreleased)
 - Fix an issue where fanning out log entries to multiple `loki.process`
   components lead to a race condition. (@tpaschalis)
 
+- Fix panix in `prometheus.operator.servicemonitors` from relabel rules without certain defaults. (@captncraig)
+
 ### Other changes
 
 - Add metrics when clustering mode is enabled. (@rfratto)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -99,7 +99,7 @@ Main (unreleased)
 - Fix an issue where fanning out log entries to multiple `loki.process`
   components lead to a race condition. (@tpaschalis)
 
-- Fix panix in `prometheus.operator.servicemonitors` from relabel rules without certain defaults. (@captncraig)
+- Fix panic in `prometheus.operator.servicemonitors` from relabel rules without certain defaults. (@captncraig)
 
 ### Other changes
 

--- a/component/prometheus/operator/configgen/config_gen.go
+++ b/component/prometheus/operator/configgen/config_gen.go
@@ -114,28 +114,18 @@ func (r *relabeler) addFromV1(cfgs ...*promopv1.RelabelConfig) (err error) {
 		for _, l := range c.SourceLabels {
 			cfg.SourceLabels = append(cfg.SourceLabels, model.LabelName(l))
 		}
-		if c.Separator != "" {
-			cfg.Separator = c.Separator
-		}
-		if c.TargetLabel != "" {
-			cfg.TargetLabel = c.TargetLabel
-		}
+		cfg.Separator = c.Separator
+		cfg.TargetLabel = c.TargetLabel
 		if c.Regex != "" {
 			cfg.Regex, err = relabel.NewRegexp(c.Regex)
 			if err != nil {
 				return err
 			}
 		}
-		if c.Modulus != 0 {
-			cfg.Modulus = c.Modulus
-		}
-		if c.Replacement != "" {
-			cfg.Replacement = c.Replacement
-		}
-		if c.Action != "" {
-			cfg.Action = relabel.Action(c.Action)
-		}
-		r.configs = append(r.configs, cfg)
+		cfg.Modulus = c.Modulus
+		cfg.Replacement = c.Replacement
+		cfg.Action = relabel.Action(c.Action)
+		r.add(cfg)
 	}
 	return nil
 }

--- a/component/prometheus/operator/configgen/config_gen_podmonitor_test.go
+++ b/component/prometheus/operator/configgen/config_gen_podmonitor_test.go
@@ -148,6 +148,12 @@ func TestGeneratePodMonitorConfig(t *testing.T) {
 						InsecureSkipVerify: true,
 					},
 				},
+				RelabelConfigs: []*promopv1.RelabelConfig{
+					{
+						SourceLabels: []promopv1.LabelName{"foo"},
+						TargetLabel:  "bar",
+					},
+				},
 			},
 			expectedRelabels: util.Untab(`
 				- source_labels: [job]
@@ -198,9 +204,10 @@ func TestGeneratePodMonitorConfig(t *testing.T) {
 				  replacement: "${1}"
 				  regex: "(.+)"
 				  target_label: job
-				
 				- target_label: endpoint
 				  replacement: metrics
+				- target_label: bar
+				  source_labels: [foo]
 			`),
 			expected: &config.ScrapeConfig{
 				JobName:         "podMonitor/operator/podmonitor/1",

--- a/component/prometheus/operator/configgen/config_gen_servicemonitor_test.go
+++ b/component/prometheus/operator/configgen/config_gen_servicemonitor_test.go
@@ -162,6 +162,12 @@ func TestGenerateServiceMonitorConfig(t *testing.T) {
 						InsecureSkipVerify: true,
 					},
 				},
+				RelabelConfigs: []*promopv1.RelabelConfig{
+					{
+						SourceLabels: []promopv1.LabelName{"foo"},
+						TargetLabel:  "bar",
+					},
+				},
 			},
 			expectedRelabels: util.Untab(`
 				- source_labels: [job]
@@ -227,6 +233,8 @@ func TestGenerateServiceMonitorConfig(t *testing.T) {
 				- target_label: endpoint
 				  replacement: metrics
 				  action: replace
+				- target_label: bar
+				  source_labels: [foo]
 			`),
 			expected: &config.ScrapeConfig{
 				JobName:         "serviceMonitor/operator/svcmonitor/1",

